LM-835 Splunk logs integration

denis-shuvalov opened the pull request:
Splunk Cloud Platform integration added:

- Collection of runtime logs of a Mendix app by Fluentbit.
- Submitting the logs to Splunk Cloud dashboard.

markbastiaans commented:
I notice that you're building FluentBit inside of the buildpack. Please move that to a dedicated Gitlab project, and upload the release artifact for that with `cdn-uploader`.

markbastiaans reviewed (changes requested):
I notice that you're building FluentBit inside of the buildpack. The buildpack project is not meant for building external dependencies.

Please move that to a dedicated Github / Gitlab project and upload the release artifact for that with `cdn-uploader`. You can also copy it to the `vendor` directory locally to test.

denis-shuvalov commented:
> See my earlier comment - please don't build artifacts for external dependencies in the buildpack project.

@markbastiaans I actually don't build Fluentbit during Stage process. Compiled Fluentbit is being uploaded to cdn and then I use it. Do you mean the scripts from dev/fluentbit? The scripts will be removed when I finish work on this PR.

pappachino commented on buildpack/telemetry/fluentbit.py:
```diff
@@ -0,0 +1,138 @@
+        "SPLUNK_TOKEN",
+    }
+
+    os_env_set = set(os.environ)
+
+    diff_envs = required_env_vars.difference(os_env_set)
+
+    if len(diff_envs) == 0:
+        return True
+    else:
+        return False
```
Isn't it better to do the `os.getenv` check and return False if any missing ? Here we're fetching all the current environment vars unnecessarily.

A simplified version
```suggestion
return all(map(os.getenv, required_env_vars))
```

denis-shuvalov replied:
It really looks much more simple, fixed

pappachino commented on etc/fluentbit/fluentbit.conf:
```diff
@@ -0,0 +1,21 @@
+[FILTER]
+    Name    lua
+    Match   *
+    script  /home/vcap/app/.local/fluentbit/redaction.lua
```
Shouldn't `redaction.lua` suffice as the lookup is relative to the config file ?

pappachino commented on etc/fluentbit/redaction.lua:
```diff
@@ -0,0 +1,35 @@
+
+    new_record["message"] = message
+
+    return 1, timestamp, new_record
```
It should be as follows as we're not modifying the timestamp. Fluent bit [docs](30d224c4-72eb-447e-b4c8-ccfe26a4a1f4)
```suggestion
    return 2, timestamp, new_record
```

denis-shuvalov replied:
fixed

pappachino commented on buildpack/telemetry/fluentbit.py:
```diff
@@ -0,0 +1,138 @@
+        )
+
+
+def is_fluentbit_used():
```
This is more of a check to ascertain if Splunk needs to be enabled, not fluent bit.
```suggestion
def is_splunk_enabled():
```

denis-shuvalov replied:
renamed

pappachino commented on etc/fluentbit/fluentbit.conf:
```diff
@@ -0,0 +1,21 @@
+[INPUT]
+    Name        tcp
+    Listen      localhost
+    Port        5170
```
This is the `LOGS_PORT` envvar

denis-shuvalov replied:
Done via en. var

denis-shuvalov commented:
> Hi @denis-shuvalov - saw your changes to the readme. Can you also update the ToC to reflect the extra section? The ToC was generated with the Markdown All-in-One plugin for VS Code (heading level H2-H3), you can either use that or if that's not an option update the ToC yourself.
> 
> Thanks in advance!

done

denis-shuvalov commented on buildpack/telemetry/fluentbit.py:
```diff
@@ -0,0 +1,138 @@
+    required_env_vars = {
+        "SPLUNK_HOST",
+        "SPLUNK_TOKEN",
+    }
```
Moved on the top

denis-shuvalov commented on README.md:
```diff
@@ -563,6 +563,31 @@ The buildpack includes a variety of telemetry agents, and can configure logging
+|-|-|-|-|
+| `SPLUNK_HOST` | `test.splunkcloud.com` | - | Host of Splunk Cloud without 'http://' |
+| `SPLUNK_PORT` | `8088` | `8088` | Port of Splunk Cloud |
+| `SPLUNK_TOKEN` * | `uuid token` | - | Token from Splunk Cloud dashboard |
```
Discussed, changed for superscript

denis-shuvalov commented on README.md:
```diff
@@ -563,6 +563,31 @@ The buildpack includes a variety of telemetry agents, and can configure logging
+### Splunk
+
+To collect Mendix Runtime logs to [Splunk Cloud Platform](https://www.splunk.com/en_us/products/splunk-cloud-platform.html), 
+[Fluentbit](https://docs.fluentbit.io/manual/) is used.
```
fixed

udh-mendix commented on buildpack/telemetry/fluentbit.py:
```diff
@@ -0,0 +1,131 @@
+    "FLUENTBIT_LOGS_PORT": os.getenv("FLUENTBIT_LOGS_PORT", default="5170"),
+}
+
+REQUIRED_FLUENTBIT_ENV_VARS = ["SPLUNK_HOST", "SPLUNK_TOKEN"]
```
@denis-shuvalov @pappachino splunk and fluentbit are not bound to each other. We'll have use cases where we use fluentbit for other integrations in the future, for which this code will need to be massively restructured. Can we handle this right away to make life easier in the future?

Thoughts?

denis-shuvalov replied:
Split the fluentbit file for splunk and fluentbit dedicated ones.

udh-mendix commented on buildpack/telemetry/fluentbit.py:
```diff
@@ -0,0 +1,131 @@
+        )
+        return
+
+    logging.info("Starting Fluent Bit...")
```
there is no log line that says splunk is enabled/configured. Isn't it confusing for customers? We mention in the docs that we use fluentbit for splunk but it should be made clear in logs as well IMO.

denis-shuvalov replied:
In splunk.py module I added a log line about splunk

pappachino commented on buildpack/start.py:
```diff
@@ -163,6 +167,7 @@ def _register_signal_handlers():
             extra_jmx_instance_config=databroker_jmx_instance_cfg,
             jmx_config_files=databroker_jmx_config_files,
         )
+        fluentbit.update_config(m2ee)
```
I notice `fluenbit.update_config()`is called twice ?

denis-shuvalov replied:
Yes, it is mistake, didn't notice it, changed

pappachino commented on buildpack/telemetry/splunk.py:
```diff
@@ -0,0 +1,41 @@
+    if not is_splunk_enabled():
+        return
+
+    logging.info("Splunk is configured.")
```
Shouldn't this line go after the `update_config()` is invoked? That's where we configure the required vars.

denis-shuvalov replied:
Log line changed

pappachino commented on buildpack/telemetry/fluentbit.py:
```diff
@@ -0,0 +1,131 @@
+    if not is_fluentbit_enabled():
+        return
+
+    logging.info("Fluent Bit is being installed...")
```
Let's add a log line after the fluent bit is successfully installed. We don't need it here.

denis-shuvalov replied:
Changed

pappachino commented on buildpack/telemetry/splunk.py:
```diff
@@ -0,0 +1,41 @@
+import logging
```
For the sake of consistency, let's add a `def run()` method in splunk.py. And it could simply log that Splunk has been successfully enabled.

denis-shuvalov replied:
added function "ready", which prints log line

pappachino commented on buildpack/telemetry/splunk.py:
```diff
@@ -0,0 +1,53 @@
+    if not is_splunk_enabled():
+        return
+
+    logging.info("Splunk is configuring...")
```
```suggestion
    logging.info("Configuring Splunk ...")
```

pappachino commented on buildpack/telemetry/fluentbit.py:
```diff
@@ -0,0 +1,133 @@
+            "Application logs will not be shipped to Fluent Bit."
+        )
+
+    splunk.ready()
```
Shouldn't this call be within the `if` block, only when fluent bit log subscriber is ready?

denis-shuvalov replied:
moved

pappachino reviewed (commented):
The current changes look good to me. Can you add an integration test for Splunk too, to check for the log lines if enabled?

denis-shuvalov commented:
> The current changes look good to me. Can you add an integration test for Splunk too, to check for the log lines if enabled?

Renamed test file, added a check for log line of Splunk

pappachino commented on tests/integration/test_splunk.py:
```diff
@@ -0,0 +1,32 @@
+        assert output is not None
+        assert str(output).find("fluent-bit") >= 0
+
+    def _test_splunk_is_configured(self):
```
Thanks, so this tests the scenario where the SPLUNK vars are enabled. Can you test the unhappy path too?

denis-shuvalov replied:
Added negative test. Added regex support methods to base module

denis-shuvalov commented on tests/integration/basetest.py:
```diff
@@ -92,6 +93,46 @@ def get_recent_logs(self):
     def await_string_in_recent_logs(self, *args, **kwargs):
         return self._runner.await_string_in_logs(*args, **kwargs)
 
+    def check_patterns_in_resent_logs(self, patterns):
```
fixed
